### PR TITLE
fix(areal): read areals from text: parse(), assign() and operator>>

### LINKED ISSUE #1454: areal: operator>> is a link error -- the friend declaration and the definition have different signatures
`std::cin >> someAreal` has never worked. It is a link error, identically before and
after #1452.

```
undefined reference to `std::istream& sw::universal::operator>><16u, 5u, unsigned short>(
    std::istream&, sw::universal::areal<16u, 5u, unsigned short>&)'
```

Reproduced byte-for-byte on `main` and on the #1452 branch:

```cpp
#include <universal/number/areal/areal.hpp>
#include <sstream>
int main() {
    std::istringstream is("1.5");
    sw::universal::areal<16,5,uint16_t> a;
    is >> a;          // links against nothing
}
```

## Two independent defects

**The declaration and the definition are different functions.** `areal_impl.hpp` declares
the friend as

```cpp
friend std::istream& operator>> (std::istream& istr, areal<nnbits,nes,nbt>& r);
```

and the definition takes a **`const`** reference:

```cpp
inline std::istream& operator>>(std::istream& istr, const areal<nnbits,nes,nbt>& v)
```

Extraction into a mutable `areal` selects the declared non-const overload, which is never
defined -- hence the link error. `-fsyntax-only` reports nothing, because a template body
is not checked until instantiated and the missing symbol is a link-time fact.

**The definition would not compile anyway.** Its body is

```cpp
istr >> v._fraction;
```

and `areal` has no `_fraction` member -- its storage is `bt _block[nrBlocks]`. So even
correcting the constness would only move the failure from link time to compile time.

## Why it is not fixed in #1452

#1452 relocates the definition from `areal_impl.hpp` to `iostream.hpp` and changes nothing
inside it; the friend declaration stays in the class. Both the declaration and the
definition are verbatim from `main`, and the link error is identical on both sides, which
the PR verifies.

Fixing it is design work, not a refactor:

- `operator<<` emits **`[d]`** for an exact areal and **`(d, dnext)`** for an uncertain
  one -- the uncertainty bit is the whole point of the type. A faithful `operator>>` has
  to decide whether it parses that interval syntax back, or only accepts a decimal and
  clears the ubit, and the two are observably different.
- There is no `parse(std::string, areal&)` to build on, unlike bfloat16, dfloat and hfloat.
- `assign(const std::string&)` is a stub that prints `"assign TBD"` and returns, so the
  obvious delegation target does not exist either.

## Suggested work

1. Decide the input grammar. Round-tripping `operator<<`'s own output is the defensible
   default: `[d]` sets the value exact, `(d, dnext)` sets the ubit.
2. Implement `parse(const std::string&, areal&)` and make `assign()` call it, mirroring
   how bfloat16/dfloat/hfloat are arranged. That also retires the `"assign TBD"` stub.
3. Correct the friend declaration and the definition to the same signature, and drop
   `_fraction`.
4. Add a regression test -- extraction is currently untested for areal, which is why a
   link error has survived this long. Related: #1453, on the manipulator templates that no
   test instantiates.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

Fixes #1454. `std::cin >> someAreal` never worked:
- **The link error:** the friend declaration took a mutable reference and the definition a `const` one, so extraction called a function that was never defined.
- **A body that couldn't compile:** the definition read a `_fraction` member `areal` does not have.
- **A stub:** `assign(std::string)` printed "assign TBD", and there was no `parse()`.

This implements the design the issue proposed: round-trip `operator<<`'s own output.

## The input forms

`parse(const std::string&, areal&)` reads the three forms an areal is written in:

| text | result |
|---|---|
| `d`, a decimal | the value truncated toward zero, with the ubit set when that loses anything. It is the same rule as conversion from `double`, applied to the exact decimal: `1.5000000000000000000001` reads as `(1.5, 1.50195)`, not `[1.5]` |
| `[d]`, the exact form `operator<<` writes | the exact areal nearest to `d`. `operator<<` prints a handful of digits, so `d` itself is usually not exact. Ties go to the even encoding |
| `(d, dnext)`, the uncertain form `operator<<` writes | the open interval between two adjacent exact values, the nearest ones to `d` and `dnext`. Either order is accepted. Endpoints that are not adjacent, such as `(1, 2)`, are rejected |

Inf and nan are accepted bare or bracketed, in the usual spellings; a nan reads as quiet.

## How

- **Exact conversion:** the decimal goes through `decimal_to_binary`, packed into a `blocktriple<fbits, MUL>` as cfloat's `parse()` does. The existing `convert(blocktriple, areal)` then truncates and sets the ubit, including for the converter's own guard and sticky bits. So `parse()` is exact at every width, including areals wider than a `double`, such as `areal<128,15>`.
- **Guards around the converter:**
  - It loses bits silently when a decimal needs more working bits than its width (#1504). So `parse()` sizes the width from the converter's own `required_working_bits()` estimate: 2048, 8192 or 32768 bits. It rejects text needing more rather than returning a wrong value.
  - A value far outside the type is saturated to `(maxpos, inf)` or underflowed to `(0, minpos)` from its order of magnitude alone, so a short `1e5000` needs no huge conversion.
- **`assign()`** calls `parse()`, and leaves the areal at zero on text that does not parse, as dfloat and bfloat16 do.
- **`operator>>`** reads the bracketed forms through their closing bracket, since `operator<<` writes a space inside the interval. Other input is one token. A failure sets `failbit` with a diagnostic on `std::cerr`, as cfloat and bfloat16 do. It is no longer a friend.
- **Layering:** `parse()` lives in the core, as hfloat's and dfloat's do, so `assign()` works from `core.hpp` alone. The core stays free of I/O headers. It grows by 2,503 preprocessed lines (51,263 to 53,766), mostly the integer core the converter uses.
- **`reset()`** now carries the same `-fipa-icf` suppression `at()` already has. Instantiating several areals over `uint8_t` in one translation unit made GCC report `-Warray-bounds` and `-Wstringop-overflow` on its guarded write, and `-fno-ipa-icf` removes both.

## Test

New: `static/range/areal/api/string_parse.cpp`, using the standard MANUAL_TESTING and REGRESSION_LEVEL skeleton.
- **An exact oracle:** every case is built from exact binary values and their exact decimal expansions, so the reference does not depend on the parser. For each exact value `v` and the next one, `w`:
  - `v`'s exact decimal reads back exactly;
  - the same decimal with a digit appended past double precision, and the exact midpoint, both read as the interval above `v`;
  - `[midpoint]` reads as the even one of `v` and `w`, `[(3v+w)/4]` as `v`, and `[(v+3w)/4]` as `w`;
  - `(v, w)` and `(w, v)` both read as the interval above `v`.
- **Round trips:** every encoding round-trips through the text `operator<<` writes, via `parse()`, `assign()` and `operator>>`.
- **Level 1** runs in 1 s:
  - every encoding of `areal<8,2>`, `<10,3>` and `<12,4>`;
  - samples of `<16,5>` and `<32,8>`, single- and multi-limb;
  - `areal<128,15>` beyond double: an exact 1 + 2^-60, 0.1 as uncertain, and saturation;
  - the reported case and a set of inputs that must be rejected.
- **Higher levels:** level 2 is exhaustive over `areal<16,5>`, level 3 samples `areal<48,11>`, and level 4 samples a million `areal<32,8>` values.
- **What is skipped:** values past `double`'s range, the top binade of an 11-bit exponent, because `operator<<` and the reference both go through `double`. No case falls in the #1503 band.

## Verification

- **This test:** levels 1-4 pass on gcc 13.3 and clang, with no warnings and 4 MB peak memory.
- **The areal suite:** all 24 tests pass on both compilers.

## Found along the way, filed separately

- **#1503:** values just above the largest finite exponent come out with the wrong sign or as NaN, in arithmetic and conversion alike. For example, `maxpos + maxpos` is -6.08e-05. Decimal input in that band inherits it until it is fixed.
- **#1504:** `decimal_to_binary` silently returns a wrong value for long decimal text. cfloat's `parse()` of the exact expansion of 2^-600 returns 0 and reports success. This parser avoids it as described above.

## Behaviour change

`BEHAVIOUR-CHANGE: areal can be read from text: operator>> (a link error before), parse(), and assign(std::string) (a stub that printed "assign TBD") accept a decimal, [d] and (d, dnext)`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01NazYbxaii2XXVzQeQKMDTs

## Summary by CodeRabbit

* **New Features**
  * Added exact decimal parsing for areal values, including bracketed intervals, uncertainty ranges, special values, and precision-aware conversion.
  * Added overflow and underflow handling with appropriate value saturation.
  * Updated stream extraction to support parsed values and bracketed intervals.

* **Bug Fixes**
  * Failed assignments now clear the destination value and report stream failures correctly.
  * Unterminated bracketed input now reports an error without modifying the destination value.

* **Tests**
  * Added comprehensive regression coverage for parsing, assignment, stream extraction, boundary values, malformed input, and round-trip conversions.

### CODE REVIEW

File: include/sw/universal/number/areal/areal_impl.hpp [hunk 1]
```diff
@@ -47,6 +47,7 @@
 #endif
 #include <universal/internal/blockbinary/blockbinary.hpp>
 #include <universal/internal/blocktriple/blocktriple.hpp>
+#include <universal/utility/decimal_to_binary.hpp>   // parse(): exact decimal text to binary (#1454)
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/areal/areal_impl.hpp:54]: Fixed in aef22051. `areal_impl.hpp` now includes `<cstddef>`, `<cstdint>` and `<string_view>` directly, for the `std::size_t`, `std::int64_t`/`std::uint64_t` and `std::string_view` that `parse()` uses. I also corrected the `<cstdio>` comment: `assign()`'s `printf` stub was its only user and is gone. The include stays, so code that reached `<cstdio>` through this header keeps working.

All 24 areal tests pass on gcc 13.3 and clang, and a translation unit that includes only `core.hpp` and calls `assign("0.1")` compiles, links and sets the ubit on both.

**Comment by Ravenwater** [comment]
@coderabbitai review

**Comment by Ravenwater** [comment]
@coderabbitai fixed the outside-diff operator>> finding in 000353f7; see commit message.

**Comment by Ravenwater** [comment]
@coderabbitai review

**Comment by Ravenwater** [comment]
@coderabbitai On your nitpick about the rejected-input list in `string_parse.cpp` (line 226 at b52c750b): no change needed. The project `.clang-format` uses `TabWidth: 4` with `UseTab: ForIndentation`. With that, the list line is 118 columns, within the 120 limit, and it is exactly what clang-format produced. `clang-format --dry-run --Werror` on those lines reports no change. It exceeds 120 only if a tab is counted as 8 columns.
